### ISSUE TITLE
Two long estimated time for simclr pretraining on ImageNet1k

### ISSUE THREAD

**Issue by SJShin-AI** [facebookresearch/vissl#372]
For pre-training simCLR with imageNet-1K, i also encountered the very long estimated time problem.

**1.  To solve the problem, i measured the time consumed for each proceduce.

- i figured out that the read_time eventually increases as the iteration goes on.
- it increases from 100 ms for each batch to 600ms.
- it seems to be a huge barrier for the pretraining on imagenet1k.**

/////////////////////////////////////////////////////////////////////////////////////////

               Timer     Host    CudaEvent
         read_sample:  133.48 ms  134.55 ms
             forward:   97.38 ms  266.44 ms
        loss_compute:    1.93 ms    1.92 ms
     loss_all_reduce:    0.06 ms    0.06 ms
            backward:  212.47 ms  530.86 ms
      optimizer_step:  358.95 ms   40.78 ms
    train_step_total:  952.28 ms  950.72 ms
INFO 2021-07-21 11:41:01,482 log_hooks.py: 277: Rank: 0; [ep: 0] iter: 10; lr: 0.60013; loss: 5.53581; btime(ms): 2276; eta: 79 days, 2:42:39; peak_mem(M): 21931; btime(5iters)(ms): 718; rolling_eta: 24 days, 23:40:19;
               Timer     Host    CudaEvent
         read_sample:  233.46 ms  234.00 ms
             forward:   53.84 ms  235.82 ms
        loss_compute:    1.31 ms    1.29 ms
     loss_all_reduce:    0.06 ms    0.06 ms
            backward:  107.39 ms  456.67 ms
      optimizer_step:  392.36 ms   43.28 ms
    train_step_total:  986.64 ms  985.90 ms
INFO 2021-07-21 11:41:11,247 log_hooks.py: 277: Rank: 0; [ep: 0] iter: 20; lr: 0.60027; loss: 5.49077; btime(ms): 1626; eta: 56 days, 12:36:40; peak_mem(M): 21931; btime(5iters)(ms): 967; rolling_eta: 33 days, 14:41:23;
               Timer     Host    CudaEvent
         read_sample:  299.77 ms  300.14 ms
             forward:   38.81 ms  225.61 ms
        loss_compute:    1.09 ms    1.07 ms
     loss_all_reduce:    0.05 ms    0.06 ms
            backward:   75.08 ms  433.37 ms
      optimizer_step:  401.15 ms   43.04 ms
    train_step_total: 1005.90 ms 1005.42 ms
INFO 2021-07-21 11:41:21,956 log_hooks.py: 277: Rank: 0; [ep: 0] iter: 30; lr: 0.60041; loss: 5.58139; btime(ms): 1441; eta: 50 days, 2:07:50; peak_mem(M): 21931; btime(5iters)(ms): 1141; rolling_eta: 39 days, 16:13:48;
               Timer     Host    CudaEvent
         read_sample:  443.04 ms  443.32 ms
             forward:   31.59 ms  220.49 ms
        loss_compute:    0.99 ms    0.97 ms
     loss_all_reduce:    0.05 ms    0.06 ms
            backward:   59.56 ms  422.02 ms
      optimizer_step:  407.15 ms   44.87 ms
    train_step_total: 1144.74 ms 1144.39 ms

////////////////////////////////////////////////////////////////////////////

**2. For the understanding of the problem, i attach the yaml file.**

/////////////////////////////////////////////////////////////////////////////

INFO 2021-07-21 11:41:01,052 log_hooks.py: 575: Train step time breakdown (rank 0):
               Timer     Host    CudaEvent
         read_sample:  133.48 ms  134.55 ms
             forward:   97.38 ms  266.44 ms
        loss_compute:    1.93 ms    1.92 ms
     loss_all_reduce:    0.06 ms    0.06 ms
            backward:  212.47 ms  530.86 ms
      optimizer_step:  358.95 ms   40.78 ms
    train_step_total:  952.28 ms  950.72 ms
INFO 2021-07-21 11:41:01,482 log_hooks.py: 277: Rank: 0; [ep: 0] iter: 10; lr: 0.60013; loss: 5.53581; btime(ms): 2276; eta: 79 days, 2:42:39; peak_mem(M): 21931; btime(5iters)(ms): 718; rolling_eta: 24 days, 23:40:19;
INFO 2021-07-21 11:41:06,411 log_hooks.py: 277: Rank: 0; [ep: 0] iter: 15; lr: 0.6002; loss: 5.50666; btime(ms): 1846; eta: 64 days, 3:55:15; peak_mem(M): 21931; btime(5iters)(ms): 985; rolling_eta: 34 days, 6:20:49;
INFO 2021-07-21 11:41:10,810 log_hooks.py: 566: Average train batch time (ms) for 20 batches: 947
INFO 2021-07-21 11:41:10,811 log_hooks.py: 575: Train step time breakdown (rank 0):
               Timer     Host    CudaEvent
         read_sample:  233.46 ms  234.00 ms
             forward:   53.84 ms  235.82 ms
        loss_compute:    1.31 ms    1.29 ms
     loss_all_reduce:    0.06 ms    0.06 ms
            backward:  107.39 ms  456.67 ms
      optimizer_step:  392.36 ms   43.28 ms
    train_step_total:  986.64 ms  985.90 ms
INFO 2021-07-21 11:41:11,247 log_hooks.py: 277: Rank: 0; [ep: 0] iter: 20; lr: 0.60027; loss: 5.49077; btime(ms): 1626; eta: 56 days, 12:36:40; peak_mem(M): 21931; btime(5iters)(ms): 967; rolling_eta: 33 days, 14:41:23;
INFO 2021-07-21 11:41:16,248 log_hooks.py: 277: Rank: 0; [ep: 0] iter: 25; lr: 0.60034; loss: 5.56199; btime(ms): 1501; eta: 52 days, 4:06:46; peak_mem(M): 21931; btime(5iters)(ms): 1000; rolling_eta: 34 days, 18:07:41;
INFO 2021-07-21 11:41:21,520 log_hooks.py: 566: Average train batch time (ms) for 30 batches: 988
INFO 2021-07-21 11:41:21,520 log_hooks.py: 575: Train step time breakdown (rank 0):
               Timer     Host    CudaEvent
         read_sample:  299.77 ms  300.14 ms
             forward:   38.81 ms  225.61 ms
        loss_compute:    1.09 ms    1.07 ms
     loss_all_reduce:    0.05 ms    0.06 ms
            backward:   75.08 ms  433.37 ms
      optimizer_step:  401.15 ms   43.04 ms
    train_step_total: 1005.90 ms 1005.42 ms
INFO 2021-07-21 11:41:21,956 log_hooks.py: 277: Rank: 0; [ep: 0] iter: 30; lr: 0.60041; loss: 5.58139; btime(ms): 1441; eta: 50 days, 2:07:50; peak_mem(M): 21931; btime(5iters)(ms): 1141; rolling_eta: 39 days, 16:13:48;
INFO 2021-07-21 11:41:31,806 log_hooks.py: 277: Rank: 0; [ep: 0] iter: 35; lr: 0.60048; loss: 5.49765; btime(ms): 1516; eta: 52 days, 17:07:53; peak_mem(M): 21931; btime(5iters)(ms): 1970; rolling_eta: 68 days, 11:08:56;
INFO 2021-07-21 11:41:36,712 log_hooks.py: 566: Average train batch time (ms) for 40 batches: 1121
INFO 2021-07-21 11:41:36,712 log_hooks.py: 575: Train step time breakdown (rank 0):
               Timer     Host    CudaEvent
         read_sample:  443.04 ms  443.32 ms
             forward:   31.59 ms  220.49 ms
        loss_compute:    0.99 ms    0.97 ms
     loss_all_reduce:    0.05 ms    0.06 ms
            backward:   59.56 ms  422.02 ms
      optimizer_step:  407.15 ms   44.87 ms
    train_step_total: 1144.74 ms 1144.39 ms
INFO 2021-07-21 11:41:37,148 log_hooks.py: 277: Rank: 0; [ep: 0] iter: 40; lr: 0.60055; loss: 5.43499; btime(ms): 1460; eta: 50 days, 18:23:16; peak_mem(M): 21931; btime(5iters)(ms): 1068; rolling_eta: 37 days, 3:11:52;
INFO 2021-07-21 11:41:42,604 log_hooks.py: 277: Rank: 0; [ep: 0] iter: 45; lr: 0.60062; loss: 5.45612; btime(ms): 1419; eta: 49 days, 8:07:46; peak_mem(M): 21931; btime(5iters)(ms): 1091; rolling_eta: 37 days, 22:04:41;
INFO 2021-07-21 11:41:47,574 log_hooks.py: 566: Average train batch time (ms) for 50 batches: 1114
INFO 2021-07-21 11:41:47,574 log_hooks.py: 575: Train step time breakdown (rank 0):
               Timer     Host    CudaEvent
         read_sample:  442.86 ms  443.09 ms
             forward:   27.13 ms  217.43 ms
        loss_compute:    0.92 ms    0.90 ms
     loss_all_reduce:    0.05 ms    0.06 ms
            backward:   50.30 ms  415.40 ms
      optimizer_step:  410.33 ms   45.41 ms
    train_step_total: 1128.88 ms 1128.60 ms

**Comment by iseessel** [comment]
Hi there,

Can you please provide the following: 

1. Amount of CPU memory and CPU power. 
2. Type of GPU used + GPU memory. 
3. Where you are training? On google-collab/SLURM cluster/cloud?

Can you check your system stats to see cpu and GPU+ memory utilization during training? Can you try increasing `NUM_DATALOADER_WORKERS`? Can you post the training stats at iter 500, usually the first couple of iterations are not indicative of the rest of the training. I will run this config on my environment and can compare.

I'm afraid that even with optimizing  `read_sample`, this training will still take a long time since it's not the bottleneck in the training step.

Unfortunately, a lot of these SSL methods are designed to use an extreme amount of compute (8gpu - 64gpu on high memory/cpu machines) with pytorch DDP and do not have reasonable 1gpu training time. If switching algorithms is possible, I would recommend looking at [dino](https://arxiv.org/pdf/2104.14294.pdf), [swav](https://arxiv.org/pdf/2006.09882.pdf), or [moco](https://openaccess.thecvf.com/content_CVPR_2020/papers/He_Momentum_Contrast_for_Unsupervised_Visual_Representation_Learning_CVPR_2020_paper.pdf), all which have VISSL implementations -- all of these are more amenable to smaller batch sizes and should improve training times (they will still take days to train though).

**Comment by iseessel** [comment]
For reference after 1 epoch here are my stats: 

```
INFO 2021-07-21 12:12:00,166 log_hooks.py: 277: Rank: 0; [ep: 0] iter: 10000; lr: 0.15; loss: 1.20127; btime(ms): 942; eta: 32 days, 15:12:59; peak_mem(M): 24739; btime(313iters)(ms): 850; rolling_eta: 29 days, 10:46:16;
INFO 2021-07-21 12:12:07,506 trainer_main.py: 202: Meters synced
INFO 2021-07-21 12:12:07,506 log_hooks.py: 566: Average train batch time (ms) for 10009 batches: 941
INFO 2021-07-21 12:12:07,506 log_hooks.py: 575: Train step time breakdown (rank 0):
               Timer     Host    CudaEvent
         read_sample:  111.05 ms  111.08 ms
             forward:   53.82 ms  247.38 ms
        loss_compute:    3.19 ms    3.17 ms
     loss_all_reduce:    0.10 ms    0.11 ms
            backward:   32.40 ms  489.35 ms
      optimizer_step:  544.73 ms   88.04 ms
    train_step_total:  940.87 ms  940.86 ms
```
